homeassistant/components/fritz/config_flow.py

```diff
@@ -0,0 +1,299 @@
+        username = user_input.get(CONF_USERNAME)
+        password = user_input.get(CONF_PASSWORD)
+
+        self.fritz_tools = await self.hass.async_add_executor_job(
+            lambda: FritzBoxTools(
+                hass=self.hass,
+                host=host,
+                port=port,
+                username=username,
+                password=password,
+            )
+        )
+        success, error = await self.hass.async_add_executor_job(self.fritz_tools.is_ok)
+
+        if not success:
+            errors["base"] = error
+            return self._show_setup_form_confirm(errors)
+
+        return self.async_create_entry(
+            title=self._name,
+            data={
+                CONF_HOST: self.fritz_tools.host,
+                CONF_PASSWORD: self.fritz_tools.password,
+                CONF_PORT: self.fritz_tools.port,
+                CONF_USERNAME: self.fritz_tools.username,
+            },
+        )
+
+    def _show_setup_form_init(self, errors=None):
+        """Show the setup form to the user."""
+        return self.async_show_form(
+            step_id="start_config",
+            data_schema=vol.Schema(
+                {
+                    vol.Optional(CONF_HOST, default=DEFAULT_HOST): str,
+                    vol.Optional(CONF_PORT, default=DEFAULT_PORT): vol.Coerce(int),
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
+            errors=errors or {},
+        )
+
+    def _show_setup_form_confirm(self, errors=None):
+        """Show the setup form to the user."""
+        return self.async_show_form(
+            step_id="confirm",
+            data_schema=vol.Schema(
+                {
+                    vol.Required(CONF_USERNAME): str,
+                    vol.Required(CONF_PASSWORD): str,
+                }
+            ),
+            description_placeholders={"name": self._name},
+            errors=errors or {},
+        )
+
+    async def async_step_user(self, user_input=None):
+        """Handle a flow initiated by the user."""
+        return await self.async_step_start_config()
+
+    async def async_step_start_config(self, user_input=None):
+        """Handle a flow start config."""
+        if user_input is None:
+            return self._show_setup_form_init()
+
+        errors = {}
+
+        host = user_input.get(CONF_HOST, DEFAULT_HOST)
+        port = user_input.get(CONF_PORT, DEFAULT_PORT)
+        username = user_input.get(CONF_USERNAME)
+        password = user_input.get(CONF_PASSWORD)
+
+        self.fritz_tools = await self.hass.async_add_executor_job(
+            lambda: FritzBoxTools(
+                hass=self.hass,
+                host=host,
+                port=port,
+                username=username,
+                password=password,
+            )
+        )
+
+        success, error = await self.hass.async_add_executor_job(self.fritz_tools.is_ok)
```

Comment:
This could be its own function since you do it many times

Reply:
Done